tools/downloader/topologies.py

```diff
@@ -0,0 +1,196 @@
+            mean = mean[mean.find('[') + 1:mean.find(']')].split(',')
+            mean = [float(val) for val in mean]
+            m.setInputMean(mean)
+
+        if '--scale_values' in self.mo_args:
+            scale = self.mo_args['--scale_values']
+            scale = 1.0 / float(scale[scale.find('[') + 1:scale.find(']')])
+            m.setInputScale(scale)
+
+        if '--input_shape' in self.mo_args:
+            shape = self.mo_args['--input_shape']
+            shape = shape[shape.find('[') + 1:shape.find(']')].split(',')
+            shape = [int(val) for val in shape]
+            if len(shape) == 4:
+                if self.framework == 'tf':  # NHWC
+                    w, h = shape[2], shape[1]
+                else:  # NCHW
+                    w, h = shape[3], shape[2]
+            m.setInputSize(w, h)
+
+        if '--reverse_input_channels' in self.mo_args:
+            m.setInputSwapRB(True)
+
+        return m
+
+
+    def getIENetwork(self):
+        '''
+        Creates openvino.inference_engine.IENetwork instance.
+        Model Optimizer is lauched to create OpenVINO Intermediate Representation (IR).
+        '''
+        from openvino.inference_engine import IENetwork
+        xmlPath, binPath = self.getIR()
+        return IENetwork(xmlPath, binPath)
+
+
+for _topology in common.load_models(argparse.Namespace(config=None)):
+    _name = _topology.name.replace('-', '_').replace('.', '_')
+    _download_dir = _cache / _topology.subdirectory
+    _mo_args = {}
+    if _topology.mo_args:
+        for _arg in _topology.mo_args:
+            _tokens = _arg.split('=')
+            _mo_args[_tokens[0]] = _tokens[1] if len(_tokens) == 2 else None
+
+    # Resolve weights and text files names.
+    if _topology.framework == 'dldt':
+        assert(len(_topology.files) >= 2), 'Expected to have at least two files in IR format'
+        _paths = []
+        # DLDT topologies usually come with multiple precision configuration files.
+        for i in range(2):
+            # Get only basename and add precision marker
+            _path = str(_topology.files[i].name)
+            _pos = _path.find('/')
+            if _path[:_pos] in common.KNOWN_PRECISIONS:
+                _path = _precision_marker + _path[_pos:]
+            _paths.append(_path)
+
+        _config_path, _model_path = _paths if _paths[0].endswith('.xml') else _paths[::-1]
+    else:
+        # Get paths from Model Optimizer arguments
+        if '--input_model' in _mo_args:
+            _model_path = _mo_args['--input_model'].replace('$dl_dir/', '')
+            _config_path = ''  # If there is a text file - we will set it next
+
+        for key in ['--input_proto', '--input_symbol']:
+            if key in _mo_args:
+                _config_path = _mo_args[key].replace('$dl_dir/', '')
+
+    _model_path = _download_dir / _model_path
+    if _config_path:
+        _config_path = _download_dir / _config_path
+
+
+    exec("""
+class {className}(Topology):
```

Comment:
I cannot respond to the original thread for this for some reason, so I'll open a new one.

It's even clearer now that the only difference between the topology classes is how they are initialized. So why do they need to be different classes? Having them be objects of the same class seems like a natural fit.

Reply:
replaced to functions. If we create objects they will start downloading.

Reply:
> If we create objects they will start downloading.

They won't if you don't implement them that way. There's no reason why the object has to start downloading the model files as soon as it's created.